tests/unit/threaded_queue_test.c

```diff
@@ -0,0 +1,171 @@
+    char *new_str3 = ThreadedQueuePop(new_queue);
+
+    assert(old_str1 == new_str1);
+    assert(old_str2 == new_str2);
+    assert(old_str3 == new_str3);
+
+    free(old_str1);
+    free(old_str2);
+    free(old_str3);
+
+    ThreadedQueueSoftDestroy(queue);
+
+    // Tests expanding the copied queue
+    ThreadedQueuePush(new_queue, xstrdup("1"));
+    ThreadedQueuePush(new_queue, xstrdup("2"));
+    ThreadedQueuePush(new_queue, xstrdup("3"));
+    ThreadedQueuePush(new_queue, xstrdup("4"));
+    ThreadedQueuePush(new_queue, xstrdup("5"));
+
+    assert_int_equal(ThreadedQueueCount(new_queue), 5);
+    assert_int_equal(ThreadedQueueCapacity(new_queue), 8);
+
+    new_str1 = ThreadedQueuePop(new_queue);
+    new_str2 = ThreadedQueuePop(new_queue);
+    new_str3 = ThreadedQueuePop(new_queue);
+    char *new_str4 = ThreadedQueuePop(new_queue);
+    char *new_str5 = ThreadedQueuePop(new_queue);
+
+    assert_int_equal(strcmp(new_str1, "1"), 0);
+    assert_int_equal(strcmp(new_str2, "2"), 0);
+    assert_int_equal(strcmp(new_str3, "3"), 0);
+    assert_int_equal(strcmp(new_str4, "4"), 0);
+    assert_int_equal(strcmp(new_str5, "5"), 0);
+
+    free(new_str1);
+    free(new_str2);
+    free(new_str3);
+    free(new_str4);
+    free(new_str5);
+
+    ThreadedQueueDestroy(new_queue);
+}
+
+static void test_push_report_count(void)
+{
+    ThreadedQueue *queue = ThreadedQueueNew(0, free);
+
+    size_t size1 = ThreadedQueuePushReportCount(queue, xstrdup("1"));
+    size_t size2 = ThreadedQueuePushReportCount(queue, xstrdup("2"));
+    size_t size3 = ThreadedQueuePushReportCount(queue, xstrdup("3"));
+    size_t size4 = ThreadedQueuePushReportCount(queue, xstrdup("4"));
+
+    assert_int_equal(size1, 1);
+    assert_int_equal(size2, 2);
+    assert_int_equal(size3, 3);
+    assert_int_equal(size4, 4);
+
+    ThreadedQueueDestroy(queue);
+}
+
+static void test_expand(void)
+{
+    ThreadedQueue *queue = ThreadedQueueNew(1, free);
+
+    ThreadedQueuePush(queue, xstrdup("spam"));
+    ThreadedQueuePush(queue, xstrdup("spam"));
+
+    char *tmp = ThreadedQueuePop(queue);
+    free(tmp);
+
+    ThreadedQueuePush(queue, xstrdup("spam"));
+    ThreadedQueuePush(queue, xstrdup("spam"));
+    ThreadedQueuePush(queue, xstrdup("spam"));
+    ThreadedQueuePush(queue, xstrdup("spam"));
+    ThreadedQueuePush(queue, xstrdup("spam"));
+
+    tmp = ThreadedQueuePop(queue);
+    free(tmp);
+    tmp = ThreadedQueuePop(queue);
+    free(tmp);
+
+    ThreadedQueuePush(queue, xstrdup("spam"));
+    ThreadedQueuePush(queue, xstrdup("spam"));
+    ThreadedQueuePush(queue, xstrdup("spam"));
+    ThreadedQueuePush(queue, xstrdup("spam"));
+    ThreadedQueuePush(queue, xstrdup("spam"));
+    ThreadedQueuePush(queue, xstrdup("spam"));
+    ThreadedQueuePush(queue, xstrdup("spam"));
+
+    assert_int_equal(ThreadedQueueCount(queue), 11);
+    assert_int_equal(ThreadedQueueCapacity(queue), 16);
+
+    ThreadedQueueDestroy(queue);
```

Comment:
Would be nice to add comments to this test that would describe how the internal array looks like (or *is supposed to look like*) at those particular stages. IOW, explanations of what the particular blocks are testing.

Reply:
I still think these comments would be useful.